File: structures/hash_table.py
fin = open('input.txt', 'r')
fout = open('output.txt', 'w')

elements = list(map(int, fin.read().split()))

class Hash_Table:
    def __init__(self):
        self.size = len(elements)
        self.keys = [0] * self.size
        self.info = [False] * self.size

    def hash_func(self, data, plus):
        return (data + plus) % self.size

    def push(self, data):
        value = self.hash_func(data, 0)

        if self.keys[value] == 0:
            self.keys[value] = data
            self.info[value] = True
        else:
            if self.keys[value] == data:
                self.keys[value] = data
            else:
                i = 1
                next_value = self.hash_func(data, i)

                while self.keys[next_value] != 0:
                    i = i + 1
                    next_value = self.hash_func(data, i)

                self.keys[next_value] = data
                self.info[next_value] = True

    def pop(self, data):
        value = self.hash_func(data, 0)

        if self.keys[value] == data:
            self.keys[value] = 0
        else:
            i = 1
            next_value = self.hash_func(data, i)

            while self.keys[next_value] != False and self.keys[next_value] != data:
                i = i + 1
                next_value = self.hash_func(data, i)

            self.keys[next_value] = 0

    def write(self):
        for i in range(len(self.keys)):
            if self.keys[i] != 0:
                fout.write(str(self.keys[i]) + ' ')


table = Hash_Table()

for i in elements:
    if i > 0:
        table.push(i)
    if i < 0:
        table.pop(-i)
    if i == 0:
        break

table.write()

File: structures/hash_table.cpp
#include <fstream>
#include <algorithm>
#include <vector>
#include <list>
#include <cmath>

#define lenght 4001

using namespace std;

ifstream fin("input.txt");
ofstream fout("output.txt");

class HashTable {
    list<int> *table;
public:
    HashTable() {
        table = new list<int>[lenght];
    }

    int hash_function(int data) {
        int result = data;
        for (int i = 1; i < data; i++) {
            result = result + pow(data % 10, i);
            data /= 10;
        }
        return abs(result % lenght);
    }

    void push(int data) {
        int key = hash_function(data);

        auto i = find(table[key].begin(), table[key].end(), data);
        if (i == table[key].end()) {
            table[key].push_back(data);
        }
    }

    void pop(int data) {
        int key = hash_function(data);

        list<int>::iterator i;
        for (i = table[key].begin(); i != table[key].end(); i++) {
            if (*i == data)
                break;
        }

        if (i != table[key].end())
            table[key].erase(i);
    }

    void print() {
        vector<int> result;
        for (int i = 0; i < lenght; i++) {
            for (auto element : table[i])
                result.push_back(element);
        }
        sort(result.begin(), result.end());
        for (int i : result) {
            fout << i << ' ';
        }
    }
};

int main() {
    HashTable table;

    int sup;
    while (fin >> sup) {
        if (sup > 0)
            table.push(sup);
        if (sup < 0)
            table.pop(-sup);
        if (sup == 0)
            break;
    }

    table.print();
}
